### ISSUE TITLE
Runtime Error: Solve(): Solution not found in Linear Regression Python Bindings

### ISSUE THREAD

**Issue by swaingotnochill** [mlpack/mlpack#2977]
#### Issue description

Running Linear Regression failed for python bindings

#### Your environment

 * version of mlpack: 3.4.2
 * operating system: linux
 * compiler: python3
 * version of dependencies (Boost/Armadillo):
 * any other environment information you think is relevant:

#### Steps to reproduce

1. Run a linear regression job with multiple features.
2. output = mlpack.linear_regression(training= train_set, training_responses = train_labels)

Colab link : https://colab.research.google.com/drive/1ywpZs5jhWi-Q-9ttVs2kZ7iY9jRMYrpK?usp=sharing

#### Expected behavior

It should execute properly.

#### Actual behavior

---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-121-77012ae5da8e> in <module>()
----> 1 output = mlpack.linear_regression(training= train_set, training_responses = train_labels)

mlpack/linear_regression.pyx in mlpack.linear_regression.linear_regression()

RuntimeError: solve(): solution not found

**Comment by swaingotnochill** [comment]
@rcurtin @zoq Do check this out... Don't know why there is such a strange behaviour.

**Comment by zoq** [comment]
In case you don't have the dataset, you can pass - https://matrix.org/_matrix/media/r0/download/matrix.org/WvrgbgzkyIMbvkxLkKKNyMrO/housing.csv to pandas, also the map is missing as well, but we can skip the step.

**Comment by rcurtin** [comment]
I can't seem to reproduce this, but to me this implies that the covariance matrix is non-invertible.  Have you tried specifying `lambda_` as a parameter?  (This will do ridge regression instead of linear regression.)  Often `lambda_` is important because if two dimensions are not linearly independent, linear regression will fail.

If you are still having the problem, can you provide an exact data matrix `train_set` and responses `train_labels` that cause the issue?

**Comment by zoq** [comment]
@swaingotnochill Can you save the two datasets from the notebook and post them here as well?

**Comment by swaingotnochill** [comment]
Sure....

**Comment by zoq** [comment]
Looks like what you posted here is the `housing.csv` file the same file I posted above, but to make it easier to reproduce, it would be great if you could post the pre-processed files as well. In this case the two files/matrices you pass to:

`output = mlpack.linear_regression(training= train_set, training_responses = train_labels)`

**Comment by swaingotnochill** [comment]
Train_set =  https://drive.google.com/file/d/1vNLZ1n0EAvvn2YFrhANyXHDWWJjRaLr9/view?usp=sharing

Train_labels = https://drive.google.com/file/d/1q8ZgRDJaUqJpFD3JqC5NtWMAcoZy2x6a/view?usp=sharing
@zoq I think you were talking about these two files.

**Comment by swaingotnochill** [comment]
> I can't seem to reproduce this, but to me this implies that the covariance matrix is non-invertible. Have you tried specifying `lambda_` as a parameter? (This will do ridge regression instead of linear regression.) Often `lambda_` is important because if two dimensions are not linearly independent, linear regression will fail.
> 
> If you are still having the problem, can you provide an exact data matrix `train_set` and responses `train_labels` that cause the issue?

@rcurtin I tried with lambda_ = 0.5, but it gave the same error. I have attached the two train_set and train_labels files. Please check with them.

**Comment by rcurtin** [comment]
Thanks!  It may be a little while until I have the time to dig in and reproduce the issue, so if you need a response quickly you might be best off digging into the code yourself to see what the issue is. :+1:

**Comment by rcurtin** [comment]
I can't reproduce this one.  Here's the transcript of my Python session, using mlpack Python bindings built from `git master`.  Can you try it on your system and tell me what you get?

```
Python 3.9.2 (default, Feb 28 2021, 17:03:44) 
[GCC 10.2.1 20210110] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import mlpack
>>> import pandas as pd
>>> import numpy as np
>>> x = np.genfromtxt('/home/ryan/Downloads/train_set.csv', delimiter=',')
>>> y = np.genfromtxt('/home/ryan/Downloads/train_labels.csv', delimiter=',')
>>> x
array([[            nan,             nan,             nan, ...,
                    nan,             nan,             nan],
       [ 9.15500000e+03, -1.18530000e+02,  3.44500000e+01, ...,
         5.74452555e+00,  1.75893992e-01,  3.13034411e+00],
       [ 1.96810000e+04, -1.21620000e+02,  3.91500000e+01, ...,
         4.74642127e+00,  1.96036191e-01,  2.38854806e+00],
       ...,
       [ 3.44500000e+03, -1.18410000e+02,  3.42500000e+01, ...,
         4.21323529e+00,  2.26003490e-01,  4.31250000e+00],
       [ 6.20300000e+03, -1.17900000e+02,  3.40800000e+01, ...,
         4.90778872e+00,  2.28201386e-01,  3.06714414e+00],
       [ 5.23000000e+03, -1.18240000e+02,  3.39300000e+01, ...,
         4.14112903e+00,  2.51217137e-01,  3.32258065e+00]])
>>> y
array([[    nan,     nan],
       [  9155., 220100.],
       [ 19681.,  74000.],
       ...,
       [  3445., 153800.],
       [  6203., 204400.],
       [  5230.,  86300.]])
>>> xx = x[1:end, 1:end]
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
NameError: name 'end' is not defined
>>> xx = x[1:-1, 1:-1]
>>> yy = y[1:-1, 1]
>>> xx
array([[-118.53      ,   34.45      ,   10.        , ...,    0.        ,
           5.74452555,    0.17589399],
       [-121.62      ,   39.15      ,   36.        , ...,    0.        ,
           4.74642127,    0.19603619],
       [-118.3       ,   33.94      ,   36.        , ...,    0.        ,
           4.3987069 ,    0.26016659],
       ...,
       [-118.02      ,   33.73      ,   24.        , ...,    0.        ,
           6.04824976,    0.17847646],
       [-118.41      ,   34.25      ,   36.        , ...,    0.        ,
           4.21323529,    0.22600349],
       [-117.9       ,   34.08      ,   32.        , ...,    0.        ,
           4.90778872,    0.22820139]])
>>> yy
array([220100.,  74000.,  99300., ..., 336900., 153800., 204400.])
>>> mlpack.linear_regression(training=xx, training_responses=yy, verbose=True)
{'output_model': <mlpack.linear_regression.LinearRegressionType object at 0x7f951a259d60>, 'output_predictions': array([], dtype=float64)}
```

Is it possible maybe you are using a different version of mlpack or something like this?

**Comment by zoq** [comment]
In the example you removed the last column as well. The first row:

`9155,-118.53,34.45,10.0,5509.0,969.0,3002.0,959.0,5.5981,0,1,0,0,0,5.744525547445256,0.17589399165002723,3.130344108446298`

the first one is the index, which we should remove but the last column is the `population`, if I remove the first and last column it works for me as well, but if I leave the last column I can see the same error.

**Comment by rcurtin** [comment]
Oh, sorry, you're right, it should be `[1:]` not `[1:-1]`.  Let me try again like that...

**Comment by rcurtin** [comment]
It seems that there are missing values in the data:

```
>>> np.argwhere(np.isnan(x)).shape
(359, 2)
```

```
>>> mlpack.linear_regression(training=xx, training_responses=yy, verbose=True, check_input_matrices=True)
[FATAL] The input training has NaN values.

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "mlpack/linear_regression.pyx", line 222, in mlpack.linear_regression.linear_regression
    IO.CheckInputMatrices()
RuntimeError: fatal error; see Log::Fatal output
```

Maybe the input dataset needs to be filtered or imputed first?

**Comment by zoq** [comment]
You are absolutely right, we are missing 207 entries for the `total_bedrooms` value:

`-118.72,34.28,17.0,3051.0,,1705.0,495.0,5.7376,218600.0,<1H OCEAN`

For the example we can just drop the column.

**Comment by rcurtin** [comment]
It could be nice to add a little block to the example to indicate that checking for missing/nan values is a good idea... you can even use the `check_input_matrices` option as a demonstration of a way to check, then remove them and run again, for instance.

**Comment by swaingotnochill** [comment]
@zoq Yes...It is just like we saw in meeting...I missed there are NaN values...This is the reason causing this error..I fixed it...I added a block which checks if there are any NaN values and if there are, we can deal according to different strategies...Sorry for this, it was due to my ignorance...

**Comment by swaingotnochill** [comment]
I am closing this issue...Again sorry for this

**Comment by rcurtin** [comment]
No worries!  Happy to help out. :)  (And also happy to check that the `check_input_matrices` option works correctly! :smile:)
